### DISCUSSION TITLE
Retraining en_core_web_lg using my own data

### DISCUSSION THREAD

**Discussion by traceymai** [explosion/spaCy#6982] [Help: Coding & Implementations]
Hi guys, since the release of spaCy v3.0 and the integration of config.cfg in all stages of model training I've been struggling to train a textcat component using a previously trained spaCy language model ("en_core_web_lg" for example). I'm not entirely sure where to declare usage of this pretrained model in the config.cfg file, previously I would load the model using spacy.load(), then add textcat pipe, do nlp.begin_training(), and so on. Would be greatly appreciated if anyone could help me out on what sections of the config file I should modify in order to account for this? Thanks in advance!

**Comment by ines** [comment]
Hi! The config system in spaCy v3 lets you "source" components from different pipelines – so instead of adding the text classifier to an existing pipeline, you can think of it more like putting together a new pipeline with some components from `en_core_web_lg` and a new component you train from scratch.

This makes it easier to re-use components from different pipelines and choose which components to update with the data and which components to "freeze" (copy as-is without updating). For example, if you want to keep the tagger and parser, copy the NER and update it with more examples, _and_ add a new text classifier, you can now express all of this in the config. See here for more details and examples: https://spacy.io/usage/training#config-components

So in your new config, you probably want to:

* set `vectors = "en_core_web_lg"` (in the section `[paths]`) to use the vectors from the `en_core_web_lg` package
* set `source = "en_core_web_lg"` for the blocks `[components.tok2vec]`, plus all other components you want to keep
* set `frozen_components` in the `[training]` block to all components you want to "freeze" and not update during training
* add a config block for the text classifier with `factory = "textcat"` or `factory = "textcat_multilabel"` (you can run [`init fill-config`](https://spacy.io/api/cli#init-fill-config) to fill in the default model, settings and hyperparameters)

**Reply by davesuketu215** [reply]
I have a use case, where the flow goes as follows:
import training text data --> automated annotation logic for training data --> feed annotated training data to blank spacy model --> feed test data to the model --> evaluate performance

I am using only the NER pipeline
Is it somehow possible to create a blank model in the config, and introduce a callback to my function for annotations. And then inturn feed this to the custom model?

**Reply by honnibal** [reply]
@davesuketu215 Perhaps what you want to do is create a custom data loader? You can make a function which streams data instead of reading it, see here: https://spacy.io/usage/training#custom-code-readers-batchers

**Reply by traceymai** [reply]
thanks so much Ines for your help! :) I managed to start the model training, but it seems after 6-8 E's I'm starting to get this error:
`PermissionError(13, 'The process cannot access the file because it is being used
by another process')
`
Any idea what might be happening? I made sure I'm not running Python from any other programme or window, and with each `spacy train` command I'd specify a different `datav3\\output{}` folder, so not too sure what's going on. I don't know if this would be the best place to ask for help regarding this, so would much appreciate any help! This is how my config.cfg looks like as reference:
`[paths]
train = "datav3//train.spacy"
dev = "datav3//test.spacy"
vectors = "en_core_web_lg"
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "en"
pipeline = ["tok2vec","textcat"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.textcat]
factory = "textcat"
threshold = 0.5

[components.textcat.model]
@architectures = "spacy.TextCatBOW.v1"
exclusive_classes = true
ngram_size = 1
no_output_layer = false
nO = null

[components.tok2vec]
source = "en_core_web_lg"


[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 2000
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
cats_score_desc = null
cats_micro_p = null
cats_micro_r = null
cats_micro_f = null
cats_macro_p = null
cats_macro_r = null
cats_macro_f = null
cats_macro_auc = null
cats_f_per_type = null
cats_macro_auc_per_type = null
cats_score = 1.0

[pretraining]

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]`
